### DISCUSSION TITLE
Inject variables into ft.app()

### DISCUSSION THREAD

**Discussion by d3jong** [flet-dev/flet#1909] [Q&A] [answered]
Any idea how to inject variables / objects into the target function? I don't prefer that that function becomes my main entrypoint for the whole application..

I want to use async functions btw.

**Comment by FeodorFitsner** [comment, marked as answer]
What are you trying to solve? You don't need injecting anything - `main` function has access to the outside scope.
In our apps we use to have `state` (or `store`) object which is a singleton in a module and which reference could be shared between multiple sessions.

**Reply by d3jong** [reply]
The goal: I have a async serialport function and some related logic functions. For that function I have two async queues, rx and tx. Next I want to let the ui communicate with those queues.

I also was thinking about a singleton, is that the best option?

**Reply by FeodorFitsner** [reply]
Singleton on a module level worked well for me. Of course, if there are multiple threads/coroutines accessing its methods modifying state make sure access is synchronized with locks/mutexes/etc.

**Reply by d3jong** [reply]
Thanks! I will try the singleton approach

**Reply by plugeit** [reply]
i have a question my app has 5 pages stored in 5 files can you please explain how to switch between the pages best way to do that i am using navigationbar

**Reply by d3jong** [reply]
@FeodorFitsner  maybe not the right discussion topic, but I'm kinda stuck on the same as @icanhackyou. 

I also have a NavigationRail, but how to change the corresponding view when clicking one of the NavigationRailDestinations? I was trying page.clear() but if I'll use that I have to "redraw" the NavRail. This is obviously not the way :sweat_smile:.

I'm creating a desktop app so Routing isn't the solution, right?

**Reply by Soif2Sang** [reply]
@icanhackyou @arnolddej take a look at routing and views. I recommend the package flet_route

**Reply by d3jong** [reply]
> @icanhackyou @arnolddej take a look at routing and views. I recommend the package flet_route

I will take a look, thanks.

Currently I'm changing the content of a row and that also does the job for me🤞
